lib/encodings/base64.fz

```diff
@@ -0,0 +1,155 @@
+              error "character '$c' at position $i is not in $encoding_name alphabet"
+            else outcome true
+      until !res.ok
+        res.err
+      else error "this should never happen"
+
+    else
+      decode data.utf8.as_array
+
+
+  # decodes a sequence of ASCII characters, decoding is strict as required by RFC 4648
+  # non alphabet characters, line breaks, missing padding cause errors
+  public decode(data array u8) =>
+
+    # determine padding length '=' is 61 in ASCII
+    pad_size :=
+      if data.count > 1 && data[data.count - 1] = 61
+        if data[data.count - 2] = 61 then 2
+        else                              1
+      else                                0
+
+    alphabet_check :=
+      for b in data
+          i:=0, i+1
+      do
+        check_char :=
+          # base64 alphabet character
+          if (b >= 65 && b <= 90) || (b >= 97 && b <= 122) || (b >= 48 && b <= 57) || (b = char_62_ascii) || (b = get_char_63)
+            outcome true
+
+          # padding character =
+          else if b = 61
+            if i < data.length - pad_size
+              error """
+                    padding character '=' not allowed within the input data, \
+                    only at the very end, as required by RFC464"""
+            else outcome true
+
+          # linebreak
+          else if b = 10 || b = 13
+              error """
+                    line breaks are not allowed within encoded data, as required by RFC464, found \
+                    {if b=10 then "LF" else "CR"} at position $i"""
+          # other non alphabet character
+          else
+            error "byte $i in input data, representing '{codepoint b.as_u32}' in ASCII, is not in $encoding_name alphabet"
+
+      until !check_char.ok
+        check_char.err
+      else outcome true
+
+    if !alphabet_check.ok       # check for non base64 characters
+      alphabet_check.err
+
+    else if !(data.length %% 4) # check input length
+      error "length of input data is not multiple of four, as required by RFC4648"
+
+    else                        # do decoding
+
+      # replace padding with zeros (character 'A' / 65 is decoded to zero)
+      input := ((data.take(data.count - pad_size)) ++ ([(u8 65)].cycle.take pad_size)).as_array
+
+      _dec_sextet(input, pos) =>
+        n := input[pos]
+
+        if n >= 65 && n <= 90       # case A-Z
+          n.as_u32 - 65
+
+        else if n >= 97 && n <= 122 # case a-z
+          n.as_u32 - 71
+
+        else if n >= 48 && n <= 57  # case 0-9
+          n.as_u32 + 4
+
+        else if n = char_62_ascii   # case +
+          u32 62
+
+        else #if n = get_char_63    # case /
+          u32 63
+
+      for
+        res list u8 := nil, res.concat_sequences [b0, b1, b2] # result: contains the decoded data at the end
+        nxt0 := 0, nxt0 + 4
+        nxt1 := 1, nxt1 + 4
+        nxt2 := 2, nxt2 + 4
+        nxt3 := 3, nxt3 + 4
+      while nxt3 < input.length
```

Comment:
I think you should be able to remove the alphabet_check above and check validity of input in this loop
`error := check_input input[pos+nxt0].bind ()->check_input input[pos+nxt1] ...)`
with check_input returning an `outcome unit`
then in else clause of loop
```
if error.is_error
  error.err
else
     # remove zero bytes caused by padding
     outcome res.take(res.count - pad_size)
```